### PR TITLE
Integrate generic sync into vcluster

### PR DESCRIPTION
**Pull request by ishankhare07**
**What issue type does this pull request address?** (keep at least one, remove the others) 
/kind enhancement
/kind feature

**What does this pull request do? Which issues does it resolve?** (use `resolves #<issue_number>` if possible) 
resolves ENG-810

**Please provide a short message that should be published in the vcluster release notes**
Add generic CRD syncing capability into vcluster core

**What else do we need to know?**

### CODE REVIEW

**Comment by ishankhare07** [comment]
configure syncing by adding the following to `values.yaml`
or to the `devspace.yaml` values
```yaml
        generic:
          clusterRole:
            extraRules:
            - apiGroups: ["apiextensions.k8s.io"]
              resources: ["customresourcedefinitions"]
              verbs: ["get", "list", "watch"]
          role:
            extraRules:
            - apiGroups: ["cert-manager.io"]
              resources: ["issuers", "certificates"]
              verbs: ["create", "delete", "patch", "update", "get", "list", "watch"]
          config: |-
            version: v1beta1
            export:  # Old from virtual cluster
            - apiVersion: cert-manager.io/v1
              kind: Issuer
            - apiVersion: cert-manager.io/v1
              kind: Certificate
              patches:
                - op: rewriteName
                  path: spec.ca.secretName
              reversePatches:
                - op: copyFromObject # Sync status back by default
                  fromPath: status
                  path: status
```

File: pkg/controllers/syncer/translator/namespaced_translator.go [hunk 1]
```diff
@@ -160,3 +166,12 @@ func (n *namespacedTranslator) TranslateMetadata(vObj client.Object) client.Obje
 func (n *namespacedTranslator) TranslateMetadataUpdate(vObj client.Object, pObj client.Object) (bool, map[string]string, map[string]string) {
 	return translate.Default.ApplyMetadataUpdate(vObj, pObj, n.syncedLabels, n.excludedAnnotations...)
 }
+
+func ConvertLabelKey(key string) string {
```
Comment (FabianKramm) [inline pkg/controllers/syncer/translator/namespaced_translator.go:170]: Please get rid of this as this couldn't be used correctly with multi-namespace syncer as we don't want to replace the labels there

File: pkg/controllers/syncer/translator/namespaced_translator.go [hunk 2]
```diff
@@ -160,3 +166,12 @@ func (n *namespacedTranslator) TranslateMetadata(vObj client.Object) client.Obje
...
+	return ConvertLabelKeyWithPrefix(LabelPrefix, key)
+}
+
+func ConvertLabelKeyWithPrefix(prefix, key string) string {
```
Comment (FabianKramm) [inline pkg/controllers/syncer/translator/namespaced_translator.go:174]: Please get rid of this as this couldn't be used correctly with multi-namespace syncer as we don't want to replace the labels there

File: pkg/controllers/syncer/translator/namespaced_translator.go [hunk 3]
```diff
@@ -17,6 +19,10 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+var (
+	LabelPrefix = "vcluster.loft.sh/label"
```
Comment (FabianKramm) [inline pkg/controllers/syncer/translator/namespaced_translator.go:23]: Please get rid of this as this couldn't be used correctly with multi-namespace syncer as we don't want to replace the labels there

File: pkg/controllers/syncer/context/context.go [hunk 4]
```diff
@@ -27,6 +27,10 @@ type RegisterContext struct {
 	Options     *controllercontext.VirtualClusterOptions
 	Controllers sets.String
 
+	// TargetNamespace is the namespace in the host cluster where
+	// vcluster should sync namespaced scope objects to
+	TargetNamespace string
```
Comment (FabianKramm) [inline pkg/controllers/syncer/context/context.go:32]: Please get rid of this as this couldn't be used correctly with multi-namespace syncer as we don't want to replace the labels there

File: pkg/controllers/generic/syncer.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
+package generic
```
Comment (FabianKramm) [inline pkg/controllers/generic/syncer.go:1]: Would be good to have a test for this for single namespace as well as multi namespace mode. Preferrably a unit test

File: pkg/config/config.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+type Import struct {
+	SyncBase `yaml:",inline" json:",inline"`
+}
+
+type SyncBase struct {
+	TypeInformation `yaml:",inline" json:",inline"`
+
+	// ID is the id of the controller. This is optional and only necessary if you have multiple syncBack or fromVirtualSyncer
```
Comment (FabianKramm) [inline pkg/config/config.go:27]: syncBack or fromVirtualSyncer shouldn't be mentioned anymore

File: pkg/config/config.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+type Export struct {
+	SyncBase `yaml:",inline" json:",inline"`
+
+	// Selector is the selector to select the objects in the host cluster.
+	// If empty will select all objects.
+	Selector *Selector `yaml:"selector,omitempty" json:"selector,omitempty"`
+
+	// // Resources to sync back to virtual cluster
+	// SyncBack []*SyncBack `yaml:"syncBack,omitempty" json:"syncBack,omitempty"`
```
Comment (FabianKramm) [inline pkg/config/config.go:48]: syncBack or fromVirtualSyncer shouldn't be mentioned anymore

File: charts/eks/values.yaml [hunk 8]
```diff
@@ -102,6 +102,11 @@ mapServices:
 hostpathMapper:
   enabled: false
 
+# generic CRD configuration
+generic:
```
Comment (FabianKramm) [inline charts/eks/values.yaml:106]: please move that to `sync.generic`

**Review by FabianKramm — CHANGES_REQUESTED** [review]

File: pkg/controllers/generic/syncer.go [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+	// 		return ctrl.Result{Requeue: true}, nil
+	// 	}
+
+	// 	f.EventRecorder().Eventf(vObj, "Warning", "SyncError", "Error syncing to virtual cluster: %v", err)
+	// 	return ctrl.Result{}, fmt.Errorf("failed to patch virtual %s %s/%s: %v", f.config.Kind, vObj.GetNamespace(), vObj.GetName(), err)
+	// } else if result == controllerutil.OperationResultUpdated || result == controllerutil.OperationResultUpdatedStatus || result == controllerutil.OperationResultUpdatedStatusOnly {
+	// 	// a change will trigger reconciliation anyway, and at that point we can make
+	// 	// a more accurate updates(reverse patches) to the virtual resource
+	// 	return ctrl.Result{}, nil
+	// }
```
Comment (matskiv) [inline pkg/controllers/generic/syncer.go:163]: Why is this section commented out?

File: pkg/controllers/generic/syncer.go [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+
+	// delete physical object because virtual one is missing
+	return syncer.DeleteObject(ctx, pObj, fmt.Sprintf("delete physical %s because virtual is missing", pObj.GetName()))
+}
+
+func (f *exporter) getControllerID() string {
+	if f.config.ID != "" {
+		return f.config.ID
+	}
+	// return plugin.GetPluginName()
```
Comment (matskiv) [inline pkg/controllers/generic/syncer.go:199]: This line should be removed?

File: pkg/controllers/register.go [hunk 11]
```diff
@@ -201,6 +211,33 @@ func RegisterControllers(ctx *context.ControllerContext, syncers []syncer.Object
...
+		return nil
+	}
+
+	configuration, err := config.Parse(c)
+	if err != nil {
+		klog.Error("error parsing the config", err.Error())
+		return err
+	}
+
+	klog.Info("generic config provided, parsed successfully")
```
Comment (matskiv) [inline pkg/controllers/register.go:231]: nit-picks: we could avoid klog usage here by wrapping the error that we return and using `loghelper.Infof` for info msgs, as we already do in another place in this file.

File: pkg/controllers/syncer/translator/namespaced_translator.go [hunk 12]
```diff
@@ -160,3 +160,12 @@ func (n *namespacedTranslator) TranslateMetadata(vObj client.Object) client.Obje
...
 }
+
+// func ConvertLabelKey(key string) string {
+// 	return ConvertLabelKeyWithPrefix(LabelPrefix, key)
+// }
+
+// func ConvertLabelKeyWithPrefix(prefix, key string) string {
+// 	digest := sha256.Sum256([]byte(key))
+// 	return translate.SafeConcatName(prefix, translate.Suffix, "x", hex.EncodeToString(digest[0:])[0:10])
+// }
```
Comment (matskiv) [inline pkg/controllers/syncer/translator/namespaced_translator.go:171]: Highlighting some commented-out code just FYI.

File: pkg/log/log.go [hunk 13]
```diff
@@ -1,5 +1,67 @@
...
+
+// func init() {
+// 	if os.Getenv("SKIP_LOG") != "true" {
+// 		if os.Getenv("DEBUG") == "true" {
+// 			ctrl.SetLogger(newLog(0))
+// 		} else {
+// 			ctrl.SetLogger(newLog(2))
+// 		}
+// 	}
+// }
```
Comment (matskiv) [inline pkg/log/log.go:20]: Highlighting some commented-out code just FYI.

File: pkg/patches/patch.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func applyPatch(obj1, obj2 *yaml.Node, patch *config.Patch, resolver NameResolver) error {
+	switch patch.Operation {
+	case config.PatchTypeRewriteName:
+		return RewriteName(obj1, patch, resolver)
+	// case config.PatchTypeRewriteLabelKey:
+	// 	return RewriteLabelKey(obj1, patch, resolver)
+	// case config.PatchTypeRewriteLabelExpressionsSelector:
+	// return RewriteLabelExpressionsSelector(obj1, patch, resolver)
```
Comment (matskiv) [inline pkg/patches/patch.go:82]: Highlighting some commented-out code just FYI.

File: pkg/patches/patch_types.go [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+// 			if err != nil {
+// 				return errors.Wrap(err, "create node")
+// 			}
+
+// 			ReplaceNode(obj1, m, newNode)
+// 		}
+// 	}
+
+// 	return nil
+// }
```
Comment (matskiv) [inline pkg/patches/patch_types.go:349]: Highlighting some commented-out code just FYI.

File: pkg/util/translate/translate.go [hunk 16]
```diff
@@ -111,6 +125,17 @@ func ResetObjectMetadata(obj metav1.Object) {
...
+func IsManaged(obj runtime.Object) bool {
+	metaAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false
+	} else if metaAccessor.GetLabels() == nil {
+		return false
+	}
+
+	return metaAccessor.GetLabels()[MarkerLabel] == Suffix
+}
```
Comment (matskiv) [inline pkg/util/translate/translate.go:137]: We already define the `IsManaged` in the Translator interface in this package, and likely this function should be removed and translate.Default.IsManaged() should be used instead.

File: pkg/controllers/generic/syncer.go [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+func (r *virtualToHostNameResolver) TranslateNameWithNamespace(name string, namespace string, regex *regexp.Regexp, _ string) (string, error) {
+	if regex != nil {
+		return patchesregex.ProcessRegex(regex, name, func(name, ns string) types.NamespacedName {
+			// if the regex match doesn't contain namespace - use the namespace set in this resolver
+			if ns == "" {
+				ns = namespace
+			}
+
+			translate.Default.PhysicalNamespace(namespace)
+			return types.NamespacedName{Namespace: r.targetNamespace, Name: translate.Default.PhysicalName(name, ns)}
```
Comment (matskiv) [inline pkg/controllers/generic/syncer.go:249]: This part seems wrong. For one, `translate.Default.PhysicalNamespace(namespace)` returns a string, but it is never captured and used.

File: pkg/controllers/generic/syncer.go [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+					Key:      r.Key,
+					Operator: r.Operator,
+					Values:   r.Values,
+				}
+			}
+		}
+		s.MatchLabels[translate.NamespaceLabel] = r.namespace
+		s.MatchLabels[translate.MarkerLabel] = translate.Suffix
+	}
+	return s, nil
```
Comment (matskiv) [inline pkg/controllers/generic/syncer.go:276]: `translate.Default.TranslateLabelSelectorCluster` should be used here^

File: pkg/controllers/generic/syncer.go [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+			s[k] = v
+		}
+		s[translate.NamespaceLabel] = r.namespace
+		s[translate.MarkerLabel] = translate.Suffix
+	}
+	return s, nil
+}
+
+func (r *virtualToHostNameResolver) TranslateNamespaceRef(namespace string) (string, error) {
+	return r.targetNamespace, nil
```
Comment (matskiv) [inline pkg/controllers/generic/syncer.go:296]: This should use `translate.Default.PhysicalNamespace`

File: pkg/controllers/generic/syncer.go [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+func (r *virtualToHostNameResolver) TranslateLabelSelector(selector map[string]string) (map[string]string, error) {
+	s := map[string]string{}
+	if selector != nil {
+		for k, v := range selector {
+			s[k] = v
+		}
+		s[translate.NamespaceLabel] = r.namespace
+		s[translate.MarkerLabel] = translate.Suffix
+	}
+	return s, nil
```
Comment (matskiv) [inline pkg/controllers/generic/syncer.go:292]: This should probably use `translate.Default.TranslateLabelSelector` - the argument types don't match, so some modifications are in order of course.

File: pkg/controllers/generic/syncer.go [hunk 21]
```diff
@@ -0,0 +1,310 @@
...
+			}
+		}
+		s.MatchLabels[translate.NamespaceLabel] = r.namespace
+		s.MatchLabels[translate.MarkerLabel] = translate.Suffix
+	}
+	return s, nil
+}
+
+func (r *virtualToHostNameResolver) TranslateLabelKey(key string) (string, error) {
+	return key, nil
```
Comment (matskiv) [inline pkg/controllers/generic/syncer.go:280]: This needs to utilize `translate.Default` to support single and multi-namespace modes. In the current implementation, it only works for multi-namespace mode.

**Review by matskiv — CHANGES_REQUESTED** [review]
My review is a mix of nit-picks, questions, and a few things that need to be fixed. Please reach out when in doubt, I think some comments might need more explanations (e.g. the ones where I refer to using `translate.Default`.

File: pkg/controllers/generic/syncer.go (hunk 17 above)
Comment (ishankhare07) [reply pkg/controllers/generic/syncer.go:249]: yea nice catch, I've changed this to 
```go
return types.NamespacedName{
  Namespace: translate.Default.PhysicalNamespace(namespace),
  Name:      translate.Default.PhysicalName(name, ns)}
```

**Review by FabianKramm — APPROVED** [review]

**Pull request merged**